### PR TITLE
Genericize wa/prepmod and use it to implement all stages of wy/wyo_appt_portal

### LINKED ISSUE #244: fetch wy/wyo_appt_portal
[learn our pipeline: fetch](https://github.com/CAVaccineInventory/vaccine-feed-ingest/wiki/Runner-pipeline-stages#fetch)

Fetch data from: https://www.wyoapptportal.org/clinic/search

Put your script in a file named: `wy/wyo_appt_portal/fetch.sh` (or `fetch.sh`, etc.)

Store it (without processing) in a new file created in the directory passed as the first argument (`sys.argv[1]`).

Check the wiki to learn more about the purpose of the fetch stage and how to get set up for development!

### Tips

1. While working on your code, run it at any point:
    ```sh
    poetry run vaccine-feed-ingest fetch wy/wyo_appt_portal
    ```

### LINKED ISSUE #245: parse wy/wyo_appt_portal
[learn our pipeline: parse](https://github.com/CAVaccineInventory/vaccine-feed-ingest/wiki/Runner-pipeline-stages#parse)

Blocked by #244

Parse existing data into [`.ndjson`](http://ndjson.org/).

Read all files in the directory passed as the second argument (`sys.argv[2]`), convert them to `.ndjson`, and then output them to new files in the directory passed as the first argument (`sys.argv[1]`).

Check the wiki to learn more about the purpose of the parse stage and how to get set up for development!

### Tips

1. Fetch data for this site before you start developing:
    ```sh
    poetry run vaccine-feed-ingest fetch wy/wyo_appt_portal
    ```

1. While working on your code, run it at any point:
    ```sh
    poetry run vaccine-feed-ingest parse wy/wyo_appt_portal
    ```

### Example
[Fetched files for `md/arcgis`](https://github.com/CAVaccineInventory/vaccine-feed-ingest-results/tree/main/md/arcgis/raw) are converted to [parsed files](https://github.com/CAVaccineInventory/vaccine-feed-ingest-results/tree/main/md/arcgis/parsed)

### LINKED ISSUE #246: normalize wy/wyo_appt_portal
[learn our pipeline: normalize](https://github.com/CAVaccineInventory/vaccine-feed-ingest/wiki/Runner-pipeline-stages#normalize)

Blocked by #245

Normalize existing [`.ndjson`](http://ndjson.org/) records into [the Vaccinate The States data schema](https://github.com/CAVaccineInventory/vaccine-feed-ingest/wiki/Normalized-Location-Schema).

Read all files in the directory passed as the second argument (`sys.argv[2]`), transform each record into the normalized schema (filling in as much detail as possible), then output them to new files in the directory passed as the first argument (`sys.argv[1]`).

Check the wiki to learn more about the purpose of the normalize stage and how to get set up for development!

### Tips

1. Fetch and parse data for this site before you start developing:
    ```sh
    poetry run vaccine-feed-ingest fetch wy/wyo_appt_portal
    poetry run vaccine-feed-ingest parse wy/wyo_appt_portal
    ```

1. While working on your code, run it at any point:
    ```sh
    poetry run vaccine-feed-ingest normalize wy/wyo_appt_portal
    ```

### Example
[Parsed files for `ak/arcgis`](https://github.com/CAVaccineInventory/vaccine-feed-ingest-results/tree/main/ak/arcgis/parsed) are converted to [normalized files](https://github.com/CAVaccineInventory/vaccine-feed-ingest-results/tree/main/ak/arcgis/normalized)

### PR DESCRIPTION
**Pull request by matthewg**
# 
# 
# 

| Key Details |
|-|
| Resolves #244 |
| Resolves #245 |
| Resolves #246 |
| State: wy |
| Site: wyo_appt_portal |

## Data sample

```json
{"id": "wyo_appt_portal:129", "name": "Museum of the Mountain Man", "address": {"street1": "Museum of the Mountain Man", "street2": null, "city": "Pinedale", "state": "WY", "zip": "82941"}, "location": null, "contact": [{"contact_type": "booking", "phone": null, "website": "https://www.wyoapptportal.org/client/registration?clinic_id=129", "email": null, "other": null}], "languages": null, "opening_dates": [{"opens": "2021-05-27", "closes": "2021-05-27"}], "opening_hours": [{"day": "Thursday", "open": "14:00", "closes": "15:30"}], "availability": {"drop_in": null, "appointments": true}, "inventory": [{"vaccine": "moderna", "supply_level": null}], "access": null, "parent_organization": null, "links": null, "notes": ["This is a drive-thru clinic at the Museum of the Mountain Man.  Please wear a mask and a short sleeve shirt.", "If you are signing up for a second dose, you must get the same vaccine brand as your first dose."], "active": null, "source": {"source": "wyo_appt_portal", "id": "Museum of the Mountain Man", "fetched_from_uri": "https://www.wyoapptportal.org/clinic/search", "fetched_at": "2021-04-28T05:45:16.759080", "published_at": null, "data": {"name": "Museum of the Mountain Man", "date": "05/27/2021", "address": "Museum of the Mountain Man, Pinedale WY, 82941", "vaccines": "Moderna COVID-19 Vaccine", "ages": "N/A", "info": "This is a drive-thru clinic at the Museum of the Mountain Man.  Please wear a mask and a short sleeve shirt.", "hours": "02:00 pm - 03:30 pm", "available": "60", "special": "If you are signing up for a second dose, you must get the same vaccine brand as your first dose.", "clinic_id": "129"}}}
```

Verified output of ak/arkgis, wa/prepmod, and wy/wyo_appt_portal

## Before Opening a PR
- [*] I tested this using the CLI (e.g., `poetry run vaccine-feed-ingest <state>/<site>`)
- [*] I ran auto-formatting: `poetry run tox -e lint-fix`

### CODE REVIEW

**Comment by eliblock** [comment]
@matthewg - I'm excited to see this PR (neat that we can begin to re-use the prepmod work done earlier this week!).

One quick note: the linter you're still running into issues with validates files against our [editorconfig](https://editorconfig.org/) setup - which configures some basic editor settings like line endings, spaces vs. tabs, etc.

Editorconfig [works automatically with many editors](https://editorconfig.org/#pre-installed), but [needs a plugin to work with some others](https://editorconfig.org/#download). If you're interested - try downloading the proper plugin for whatever editor you use. After that the configuration will be automatically enforce on file-save, and you won't have issues with that linter!

**Comment by matthewg** [comment]
@eliblock Thanks, I'll take a look at editorconfig! (And let me also check whether the "did you run this command before submitting the PR?" in the PR template can be expanded to include the SuperLint checks...)

Anyway, lint errors should be fixed now.

**Comment by tmb** [comment]
This looks fantastic! A ton of states are doing scheduling through prepmod so this will absolutely be helpful as we go forward :)

**Comment by matthewg** [comment]
@tmb Thanks! I stand on the shoulders of @jakepusateri .

**Comment by matthewg** [comment]
Rebased and resolved conflicts

**Comment by eliblock** [comment]
@matthewg I know the lint output can be a bit tricky to read, so I wanted to pull out the error message:
```
vaccine_feed_ingest/runners/_shared/normalize.py:16:1: F811 redefinition of unused 'schema' from line 13
```

It looks like line `16` in that file can be deleted.

**Comment by matthewg** [comment]
> @matthewg I know the lint output can be a bit tricky to read, so I wanted to pull out the error message:
> 
> ```
> vaccine_feed_ingest/runners/_shared/normalize.py:16:1: F811 redefinition of unused 'schema' from line 13
> ```
> 
> It looks like line `16` in that file can be deleted.

Done, thanks!

**Comment by eliblock** [comment]
Hi! We've added significant validation to the schema, and want to make sure this generic-ified normalizer takes advantage of it (that's why I've held off merging this PR).

Could you pull in the latest `main` and run your code on top of it? Full details on what changed and why are in #278

**Comment by matthewg** [comment]
@eliblock @tmb -- I see two checks are still in "expected" state, is there something I need to do?

* Run all-stages for modified runners
* You must rebase and install to get new schema v1.0.0

**Comment by eliblock** [comment]
@matthewg sorry about that! TL;DR: rebase and then re-run `poetry install --extras lint`.

We added significantly more validation to the schema (a lot more detail in #278), and a CI job which runs the pipelines that you modify. Collectively, these should enable us to _approve and merge PRs faster_ (because there will be stronger up-front guarantees that things work!) and will _enable us to share better-normalized data with our partners and the website_.

Because these updates are important to our data quality, we marked them as required even though some branches don't have them yet. You can pick them up by rebasing, re-running the poetry install command, making sure your code passes the new validations, and then pushing here!

**Comment by matthewg** [comment]
@eliblock Rebased, and fixed resulting validation errors.

**Comment by tmb** [comment]
Something seems to be funky with the parsing stage — on `wa/prepmod` only one site is being pulled per page.

**Comment by matthewg** [comment]
@tmb Oops. Fixed, nice catch!

**Comment by matthewg** [comment]
@eliblock @tmb Friendly ping. If this PR is no longer relevant, let me know and I'll close it. If it *is* still relevant -- is there a way to tighten the:

1. Someone approves running workflows.
2. Workflows detect a problem / there's a merge conflict / PrepMod changed their URL schema / ...
3. I fix the problem.
4. GOTO 1

cycle we seem to be caught in here?

**Comment by bryanculbertson** [comment]
Sorry about the review blockage. I approved it to run. I can review later tonight if Eli or tmb hasnt gotten to it yet

File: vaccine_feed_ingest/runners/_shared/fetch.py
Comment (bryanculbertson) [inline vaccine_feed_ingest/runners/_shared/fetch.py:64]: minor: look into using `urllib.parse.urljoin` for URL munging. It will help with things like avoiding making a url with `//` in the path.

https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urljoin

File: vaccine_feed_ingest/runners/_shared/fetch.py
Comment (bryanculbertson) [inline vaccine_feed_ingest/runners/_shared/fetch.py:67]: minor: look into using `urllib.parse.urlencode` for generating query args. It will help with things where the values have a `&` in them and such

https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urlencode`

File: vaccine_feed_ingest/runners/_shared/fetch.py
Comment (bryanculbertson) [inline vaccine_feed_ingest/runners/_shared/fetch.py:74]: minor: f-strings are sightly easier to read for inline formatting e.g. f"{page}.html"

File: vaccine_feed_ingest/runners/_shared/normalize.py
Comment (bryanculbertson) [inline vaccine_feed_ingest/runners/_shared/normalize.py:42]: minor: use `urllib.parse.urljoin`

File: vaccine_feed_ingest/runners/_shared/parse.py
Comment (bryanculbertson) [inline vaccine_feed_ingest/runners/_shared/parse.py:74]: Can you catch a specific exception here? It is especially dangerous to catch all errors since you are ignoring the error without logging it so you may miss an important error here.

File: vaccine_feed_ingest/runners/_shared/parse.py
Comment (bryanculbertson) [inline vaccine_feed_ingest/runners/_shared/parse.py:75]: Is it it worth logging this case, or is it expect error that doesn't need to be logged?

**Review by bryanculbertson — CHANGES_REQUESTED** [review]
A couple minor things which you can feel free to address or ignore, and one major thing that I would like fixed if possible before merging.

I am concerned about the one place you are catching all Exceptions and ignoring them, can you address that part by either catching the specific error, or adding a comment why you have to catch all exceptions?

**Comment by bryanculbertson** [comment]
@matthewg Please ping my with `@bryanculbertson` when this is ready for re-review, and I will attempt a quick turn-around so we can get this in.

File: vaccine_feed_ingest/runners/_shared/fetch.py
Comment (matthewg) [reply vaccine_feed_ingest/runners/_shared/fetch.py]: Done

File: vaccine_feed_ingest/runners/_shared/fetch.py
Comment (matthewg) [reply vaccine_feed_ingest/runners/_shared/fetch.py]: Done

File: vaccine_feed_ingest/runners/_shared/fetch.py
Comment (matthewg) [reply vaccine_feed_ingest/runners/_shared/fetch.py]: Done

File: vaccine_feed_ingest/runners/_shared/normalize.py
Comment (matthewg) [reply vaccine_feed_ingest/runners/_shared/normalize.py]: Done

File: vaccine_feed_ingest/runners/_shared/parse.py
Comment (matthewg) [reply vaccine_feed_ingest/runners/_shared/parse.py]: Narrowed the scope of the try block, and now catching `IndexError`.

**Comment by matthewg** [comment]
Thanks @bryanculbertson ! I have:

* Rebased
* Addressed code-review feedback
* Ran lint-fix
* Ran fetch/parse/normalize on both wa/prepmod and wy/wyo_appt_portal and eyeballed the output

File: vaccine_feed_ingest/runners/_shared/parse.py
Comment (bryanculbertson) [reply vaccine_feed_ingest/runners/_shared/parse.py:70]: Oh, I see! This is erroring if `row_matches` is an empty list, or contents is empty.

I am going to approve and merge, but if you have a chance to get back to this an improvement would be to change this to test for that case directly

e.g.
```
if row_matches and row_matches[-1].contents:
    content = row_matches[-1].contents[offset]
else:
    return ""
```

**Review by bryanculbertson — APPROVED** [review]

**Comment by bryanculbertson** [comment]
@matthewg Thank you for pushing this through and being patient with us. Feel free to `@bryanculbertson` if a PR gets stuck like this again.

**Pull request merged**
